beginner_source/blitz/autograd_tutorial.py

```diff
@@ -1,198 +1,276 @@
-# could not compute the full Jacobian directly, but if we just
-# want the vector-Jacobian product, simply pass the vector to
-# ``backward`` as argument:
-v = torch.tensor([0.1, 1.0, 0.0001], dtype=torch.float)
-y.backward(v)
-
-print(x.grad)
-
-###############################################################
-# You can also stop autograd from tracking history on Tensors
-# with ``.requires_grad=True`` either by wrapping the code block in
-# ``with torch.no_grad():``
-print(x.requires_grad)
-print((x ** 2).requires_grad)
-
-with torch.no_grad():
-	print((x ** 2).requires_grad)
-
-###############################################################
-# Or by using ``.detach()`` to get a new Tensor with the same
-# content but that does not require gradients:
-print(x.requires_grad)
-y = x.detach()
-print(y.requires_grad)
-print(x.eq(y).all())
-
-
-###############################################################
-# **Read Later:**
-#
-# Document about ``autograd.Function`` is at
-# https://pytorch.org/docs/stable/autograd.html#function
+#
+#
+#      J^{T}\cdot \vec{v}=\left(\begin{array}{ccc}
+#       \frac{\partial y_{1}}{\partial x_{1}} & \cdots & \frac{\partial y_{m}}{\partial x_{1}}\\
+#       \vdots & \ddots & \vdots\\
+#       \frac{\partial y_{1}}{\partial x_{n}} & \cdots & \frac{\partial y_{m}}{\partial x_{n}}
+#       \end{array}\right)\left(\begin{array}{c}
+#       \frac{\partial l}{\partial y_{1}}\\
+#       \vdots\\
+#       \frac{\partial l}{\partial y_{m}}
+#       \end{array}\right)=\left(\begin{array}{c}
+#       \frac{\partial l}{\partial x_{1}}\\
+#       \vdots\\
+#       \frac{\partial l}{\partial x_{n}}
+#       \end{array}\right)
+#
+# This characteristic of vector-Jacobian product is what we use in the above example;
+# ``external_grad`` represents :math:`\vec{v}`.
+#
+
+
+
+######################################################################
+# Computational Graph
+# ~~~~~~~~~~~~~~~~~~~
+#
+# Conceptually, autograd keeps a record of data (tensors) & all executed
+# operations (along with the resulting new tensors) in a directed acyclic
+# graph (DAG) consisting of
+# `Function <https://pytorch.org/docs/stable/autograd.html#torch.autograd.Function>`__
+# objects. In this DAG, leaves are the input tensors, roots are the output
+# tensors. By tracing this graph from roots to leaves, you can
+# automatically compute the gradients using the chain rule.
+#
+# In a forward pass, autograd does two things simultaneously: \* run the
+# requested operation to compute a resulting tensor, and \* maintain the
+# operation’s *gradient function* in the DAG. This is stored in the
+# resulting tensor’s .\ ``grad_fn`` attribute.
+#
+# The backward pass kicks off when ``.backward()`` is called on the DAG
+# root. Autograd then \* computes the gradients from each ``.grad_fn``, \*
+# accumulates them in the respective tensor’s ``.grad`` attribute, and \*
+# using the chain rule, propagates all the way to the leaf tensors.
+#
```

Comment:
Are there any pictures or graphics that would work to show this?

Reply:
Added an image of the DAG. Would you say the description is accurate and informative? cc: @soumith